### PR TITLE
Rename trait type names from I$Name to $Name

### LINKED ISSUE #727: [style] Rename trait type names from `I$Name` to `$Name`
**Summary**

E.g: rename `ITable` to `Table`. There is no point to add `I` prefix before a trait.

In rust, A trait type has two usages: 

1. Template argument constrait: like `Where T: Table`, we can know Table is a trait as soon as we see it. 
2. Trait objects: like `Box<dyn Table>`, dyn trait requires exiplit dyn keyword, there is also no confusion at all.

For the reason above, trait named with `I$Name` is meaningless.

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://datafuse.rs/policies/cla/

## Summary

Rename trait type names from I$Name to $Name

## Changelog

- Renames :
  - `ITable` to `Table`
  - `IDatabase` to `Database`

- Removes `IDataSource`, use `struct DataSource` directly

- And relevant code 

## Related Issues

Fixes #727

## Test Plan

No extra ut/stateless_test

### CODE REVIEW

**Comment by sundy-li** [comment]
Try using `pub trait I.*` to find the traits starting with I.


image

**Review by BohuTANG — APPROVED** [review]
LGTM

**Comment by BohuTANG** [comment]
> Try using `pub trait I.*` to find the traits started with I.
> 
> image

Great, bot not work again

**Comment by dantengsky** [comment]
> Try using `pub trait I.*` to find the traits started with I.
> 
> image

got it

**Comment by dantengsky** [comment]
> Try using `pub trait I.*` to find the traits started with I.
> 
> image


@sundy-li  pls take another look.

File: fusequery/query/src/sql/plan_parser.rs [hunk 1]
```diff
@@ -563,18 +563,18 @@ impl PlanParser {
     }
 
     fn create_relation(&self, relation: &sqlparser::ast::TableFactor) -> Result<PlanNode> {
-        use sqlparser::ast::TableFactor::*;
+        use sqlparser::ast::TableFactor as Ast;
 
         match relation {
-            Table { name, args, .. } => {
+            Ast::Table { name, args, .. } => {
```
Comment (drmingdrmer) [inline fusequery/query/src/sql/plan_parser.rs:569]: Why not just `TableFactor::Table`?

Comment (dantengsky) [reply fusequery/query/src/sql/plan_parser.rs:569]: yeah, `TableFactor::Table` will be better

Comment (dantengsky) [reply fusequery/query/src/sql/plan_parser.rs]: PTAL

**Comment by drmingdrmer** [comment]
The converted trait name feels weird.

Most trait names in rust are verbs. Intuitively, a verb defines some **action**, such as `Index` or `Drop`.
But a noun such as `Table`, does not feel like action. In which case, `ITable` is better: a collection of actions related to `Table`.

A trait `Table` tells a reader it is a `struct` at the first sight. Then the reader realized it is actually a  trait by re-examining its context.

People read source code by intuition, not logic(that is why good naming and indentation are important).

**Comment by dantengsky** [comment]
> The converted trait name feels weird.
> 
> Most trait names in rust are verbs. Intuitively, a verb defines some **action**, such as `Index` or `Drop`.
> But a noun such as `Table`, does not feel like action. In which case, `ITable` is better: a collection of actions related to `Table`.
> 
> A trait `Table` tells a reader it is a `struct` at the first sight. Then the reader realized it is actually a trait by re-examining its context.
> 
> People read source code by intuition, not logic(that is why good naming and indentation are important).

I have the same feeling when I first saw `Stream` and `StreamExt`. Their names give me the intuition that `Stream` is a struct, and `StreamExt` is a trait (like `File` and `FileExt`).

IMHO, it is a controversial topic,  see https://github.com/rust-lang/api-guidelines/discussions/28#:~:text=Instead%2C%20trait%20names%20adhere%20to,%2C%20ToOwned%20%2C%20AsRef%20%2C%20Extend%20.

**Comment by dantengsky** [comment]
Bot is too high (again) :laughing: 
I am turning this PR into the draft state.

**Comment by drmingdrmer** [comment]
> > The converted trait name feels weird.
> > Most trait names in rust are verbs. Intuitively, a verb defines some **action**, such as `Index` or `Drop`.
> > But a noun such as `Table`, does not feel like action. In which case, `ITable` is better: a collection of actions related to `Table`.
> > A trait `Table` tells a reader it is a `struct` at the first sight. Then the reader realized it is actually a trait by re-examining its context.
> > People read source code by intuition, not logic(that is why good naming and indentation are important).
> 
> I have the same feeling when I first saw `Stream` and `StreamExt`. Their names give me the intuition that `Stream` is a struct, and `StreamExt` is a trait (like `File` and `FileExt`).
> 
> IMHO, it is a controversial topic, see [rust-lang/api-guidelines#28](https://github.com/rust-lang/api-guidelines/discussions/28).

You are right. There is not an answer to which convention is better.
No matter what it is, it won't be a problem when people got used to it.
Think of the QWERTY keyboard layout, which we've all adapted to. 🤔

**Comment by dantengsky** [comment]
couldn't agree more :D

**Comment by BohuTANG** [comment]
After your discussion, I learned a lot and even suspect that the word 'trait' is not used well. It confused me some time when I begin to study rust.

There was another discussion:
https://internals.rust-lang.org/t/naming-convention-for-traits/1796

**Comment by BohuTANG** [comment]
See Monday, just woke up to pee and note the notification, night!

**Comment by TennyZhuang** [comment]
so much changes

LGTM

**Comment by BohuTANG** [comment]
@TennyZhuang 
After the change, we found that it does not quite meet the expectations (our expectations are readability), and this task is temporarily put to lowered in priority.
FYI

**Comment by drmingdrmer** [comment]
> See Monday, just woke up to pee and note the notification, night!

This must be the most constructive pee I've ever seen.

**Comment by dantengsky** [comment]
> @TennyZhuang
> After the change, we found that it does not quite meet the expectations (our expectations are readability), and this task is temporarily put to lowered in priority.
> FYI


@TennyZhuang 

:broken_heart:  I am closing this PR (to keep the PR tab page clean).  


pls feel free to reopen it when the time is right

**Comment by dantengsky** [comment]
@TennyZhuang  by "when the time is right", I mean anytime.. :D

**Comment by TennyZhuang** [comment]
Another discussion to #744

**Pull request closed without merging**